### ISSUE TITLE
Further improvement to "hide flyout" addon

### ISSUE THREAD

**Issue by sfederici** [ScratchAddons/ScratchAddons#7196]
### Why this would be helpful

When the flyout is setup to be automatically open when the mouse is hovering the categories, usually the user move the mouse on the category they want to get the block from. As using a sequence of blocks from the same category is very rare, the user has to click every time on the category they want to drag a new block from.

### My suggestion

When the flyout is setup to be automatically open when the mouse is hovering the categories, it would be quicker to show in the palette directly the blocks of the category the mouse is hovering at that very moment, so to spare a click on the category every time. 

### Possible alternatives

None

**Comment by DNin01** [comment]
Kind of like this?

https://github.com/ScratchAddons/ScratchAddons/assets/106490990/cc7e0414-eaa5-43af-93f7-d6ed58e53888

**Comment by DNin01** [comment]
By the way, categories can be selected by dispatching a `mouseup` event on them. Here's the script I used in the demonstration:
```js
document.querySelectorAll(".scratchCategoryMenuItem").forEach(elem =>
  elem.addEventListener("mouseenter", () => elem.dispatchEvent(new MouseEvent("mouseup")))
);
```

**Comment by sfederici** [comment]
> Kind of like this?

Exactly!!!! 

`document.querySelectorAll(".scratchCategoryMenuItem").forEach(elem =>
  elem.addEventListener("mouseenter", () => elem.dispatchEvent(new MouseEvent("mouseup")))
);`

Is this the only necessary modification? I mean, even wiring it to me its fine. If so, what is the script that must be modified?

**Comment by WorldLanguages** [comment]
Should it continue to select every category that is hovered? Or should it be just for the first one that is hovered after the mouse moves from the code area to the categories?

**Comment by sfederici** [comment]
I guess the behavior shown by @DNin01 is the perfect one.

**Comment by DNin01** [comment]
I wonder if this behavior could be a separate addon, so even if you don't use `hide-flyout`, you can still choose to select categories on hover! We could link them together with a "related addons" feature (#5937).

**Comment by Samq64** [comment]
How would we make this compatible with `columns`?

**Comment by DNin01** [comment]
> How would we make this compatible with `columns`?

If `hide-flyout` and `columns` are already compatible, it shouldn't be too hard. Are they?

**Comment by Samq64** [comment]
> > How would we make this compatible with `columns`?
> 
> If `hide-flyout` and `columns` are already compatible, it shouldn't be too hard. Are they?

The way the categories are layed out with `columns` makes it pretty much impossible to select a category without accidentally hovering over the ones below to actually get to the blocks.

**Comment by DNin01** [comment]
> > > How would we make this compatible with `columns`?
> > 
> > 
> > If `hide-flyout` and `columns` are already compatible, it shouldn't be too hard. Are they?
> 
> The way the categories are layed out with `columns` makes it pretty much impossible to select a category without accidentally hovering over the ones below to actually get to the blocks.

I suppose we'll have to add a delay, then.

**Comment by sfederici** [comment]
> By the way, categories can be selected by dispatching a `mouseup` event on them. Here's the script I used in the demonstration:
> 
> ```js
> document.querySelectorAll(".scratchCategoryMenuItem").forEach(elem =>
>   elem.addEventListener("mouseenter", () => elem.dispatchEvent(new MouseEvent("mouseup")))
> );
> ```

Please, @DNin01 , let me know where I should change the addon sourcecode to get this result (or the Scratch source code; I'm using a custom version of Scratch, so to me it is prefectly fine to change the Scratch source code). Thanks in advance

**Comment by WorldLanguages** [comment]
@sfederici That code will stop working after switching tabs. For example, after going to the costumes tab, and back to the code tab.

**Comment by WorldLanguages** [comment]
> I wonder if this behavior could be a separate addon, so even if you don't use `hide-flyout`, you can still choose to select categories on hover! We could link them together with a "related addons" feature (#5937).

That's what I was thinking.



> The way the categories are layed out with `columns` makes it pretty much impossible to select a category without accidentally hovering over the ones below to actually get to the blocks.

Good catch.

**Comment by DNin01** [comment]
> > By the way, categories can be selected by dispatching a `mouseup` event on them. Here's the script I used in the demonstration:
> > ```js
> > document.querySelectorAll(".scratchCategoryMenuItem").forEach(elem =>
> >   elem.addEventListener("mouseenter", () => elem.dispatchEvent(new MouseEvent("mouseup")))
> > );
> > ```
> 
> Please, @DNin01 , let me know where I should change the addon sourcecode to get this result (or the Scratch source code; I'm using a custom version of Scratch, so to me it is prefectly fine to change the Scratch source code). Thanks in advance

I've been working on a prototype of this for the addon, and right now I just inserted the following block of code at line 315.
```diff
@@ -312,6 +312,13 @@ export default async function ({ addon, console, msg }) {
         if (!addon.self.disabled && (toggleSetting === "hover" || toggleSetting === "cathover")) onmouseleave(e);
       };
     }
+    toolbox.addEventListener("mouseover", (e) => {
+      if (addon.self.disabled) return;
+      const element = e.target.closest(".scratchCategoryMenuItem");
+      if (element && !element.classList.contains("categorySelected") && getToggleSetting() === "cathover") {
+        element.dispatchEvent(new MouseEvent("mouseup"));
+      }
+    });
     placeHolderDiv.onmouseenter = (e) => {
       if (!addon.self.disabled && getToggleSetting() === "hover") onmouseenter(e);
     };
```

**Comment by TheColaber** [comment]
@DNin01, your implementation works, but the best way to go about it is to use the scratch provided blockly methods so that it works even when you switch between tabs and that the code is less "hacky".

**Comment by TheColaber** [comment]
Here is some code that should do the trick, you can run it your console to try.
```js
let Blockly = await scratchAddons.eventTargets.tab[0].traps.getBlockly()
let oldCreateDom = Blockly.Toolbox.Category.prototype.createDom;
Blockly.Toolbox.Category.prototype.createDom = function() {
    oldCreateDom.call(this)
    let toolbox = this.parent_.parent_;
    Blockly.bindEvent_(this.item_, 'mouseover', toolbox, toolbox.setSelectedItemFactory(this));
}
let toolbox = Blockly.getMainWorkspace().getToolbox()
let existingCategories = toolbox.categoryMenu_.categories_
for (let category of existingCategories) {
    Blockly.bindEvent_(category.item_, 'mouseover', toolbox, toolbox.setSelectedItemFactory(category));
}
```

**Comment by GarboMuffin** [comment]
something you can try in TurboWarp:
```js
let oldCreateDom = ScratchBlocks.Toolbox.Category.prototype.createDom;
ScratchBlocks.Toolbox.Category.prototype.createDom = function() {
    oldCreateDom.call(this)
    let toolbox = this.parent_.parent_;
    ScratchBlocks.bindEvent_(this.item_, 'mouseover', toolbox, toolbox.setSelectedItemFactory(this));
}
let toolbox = ScratchBlocks.getMainWorkspace().getToolbox()
let existingCategories = toolbox.categoryMenu_.categories_
for (let category of existingCategories) {
    ScratchBlocks.bindEvent_(category.item_, 'mouseover', toolbox, toolbox.setSelectedItemFactory(category));
}
```

**Comment by sfederici** [comment]
> something you can try in TurboWarp:

Yes!!! Excellent! It works like a charm. Can I add this code to the "usual" index.jsx file?

**Comment by GarboMuffin** [comment]
was hoping someone would throw this into an addon and i could pull that but it appears that has not happened yet

**Comment by sfederici** [comment]
I try to add to code at the end of the userscript.js in the hide-flyout addon but it didn't work either. I guess I will have learn more about how Scratch's source code is organized and executed.

**Comment by sfederici** [comment]
> Here is some code that should do the trick, you can run it your console to try.
> 
> ```js
> let Blockly = await scratchAddons.eventTargets.tab[0].traps.getBlockly()
> let oldCreateDom = Blockly.Toolbox.Category.prototype.createDom;
> ...
> ```

I tested this code for a few days. I it really great but I think we need an -even very small- delay before the category blocks are updated in the flyout. Indeed, if you move the mouse, e.g., onto the Control category, then move the mouse diagonally to get to the "wait seconds" block (as in the picture), you have to pass through the Events category button. This will make the blocks shown in the palette change before you get to the "wait seconds" block.

<img width="210" alt="immagine" src="https://github.com/ScratchAddons/ScratchAddons/assets/2651868/74a90a9c-8efb-4b98-b34e-0218d8857a40">

**Comment by DNin01** [comment]
> but I think we need an -even very small- delay before the category blocks are updated in the flyout.

That would also solve the `columns` addon problem, theoretically!
